infra/build/functions/ci.py

```diff
@@ -0,0 +1,260 @@
+# Copyright 2022 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+################################################################################
+"""Tool for testing changes to base-images in OSS-Fuzz. This script builds test
+versions of all base images and the builds projects using those test images."""
+import argparse
+import collections
+import logging
+import sys
+
+from googleapiclient.discovery import build as cloud_build
+import oauth2client.client
+import requests
+
+import build_and_push_test_images
+import build_and_run_coverage
+import build_project
+
+IMAGE_PROJECT = 'oss-fuzz'
+BASE_IMAGES_PROJECT = 'oss-fuzz-base'
+TEST_IMAGE_SUFFIX = 'testing'
+FINISHED_BUILD_STATUSES = ('SUCCESS', 'FAILURE', 'TIMEOUT', 'CANCELLED',
+                           'EXPIRED')
+
+
+class ProjectStatus:
+  """Class that holds info about project builds."""
+
+  def __init__(self, name):
+    self.name = name
+    self.production_build = {'coverage': None, 'fuzzing': None}
```

Comment:
I think we can make this code a lot cleaner.

A big part of this is the way this class is structured, it's is responsible for a lot of different things:
- Storing results for **both** coverage and fuzzing builds.
- Storing production build results. This is never used again after computing the list of projects to build.
- Storing the build result, and build ID mappings.
- Storing the finished status of the test builds.

Let's try to simplify all of these in a few ways:

1. Only handle one build type at a time in all of these functions and this class. There are many other parts of the code with `if build_type == 'coverage': then blah else blah`. This would eliminate that and enable cleaner code reuse and extensibility in the future. e.g. what happens when we want to add introspector builds to the mix? 

2. Remove the production build results from this class. We just need to compute this once when computing the list of projects to build.

3. Remove finished status/build results from this class and handle that separately. 

This would look something like:

```python

class BuildConfiguration:
  def __init__(self, type, get_build_steps, status_json_name):
    ...


class ProjectTestBuild:  # Simplified/renamed from ProjectStatus
  def __init__(self, name, build_id):
    self.name = name
    self.build_id = build_id

BUILD_CONFIGS = [
  BuildConfiguration('fuzzing', get_build_steps, 'status.json'),
  BuildConfiguration('coverage', get_covarge_build_steps, 'status-coverage.json'),
]

def run_test_builds(build_config) -> [ProjectTestBuild]:
  project  = get_projects(build_config)  # Filter on successful production builds.
  # ... create builds and return IDs


def wait_and_check_builds(all_test_builds):
  for type, builds in all_test_builds.items():
    results = {}  # map of project -> build status. Only set if the build is done. 
    while True:
      if check_builds_done(builds, results)  # this func knows it's done if len(results) == len(builds)
        # maybe early exit here if any failures already detected over threshold?
        break

      time.sleep(..)

  # return if failures over threshold. 
    

def main():
  all_test_builds = {}  # map of type -> [ProjectTestBuild]
  for build_config in BUILD_CONFIGS:
    all_test_builds[build_config.type] = run_test_builds(build_config)

  return wait_and_check_status(all_test_builds)
```

WDYT?